File: src/com/hughes/android/dictionary/PreferenceFragment.java
package com.hughes.android.dictionary;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import android.text.InputType;
import android.widget.EditText;

import java.util.List;

public class PreferenceFragment extends PreferenceFragmentCompat {
    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        if (preference.getKey().equals(getResources().getString(R.string.quickdicDirectoryKey))) {
            final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
            String current = prefs.getString(getResources().getString(R.string.quickdicDirectoryKey), "");
            EditText t = new EditText(getActivity());
            t.setText(current);
            t.setInputType(InputType.TYPE_CLASS_TEXT);
            AlertDialog.Builder b = new AlertDialog.Builder(getActivity())
                    .setTitle(getActivity().getString(R.string.quickdicDirectoryTitle))
                    .setView(t)
                    .setNegativeButton(getString(android.R.string.cancel), null)
                    .setPositiveButton(getString(android.R.string.ok), (dialogInterface, i) -> {
                        final SharedPreferences prefs1 = PreferenceManager.getDefaultSharedPreferences(getContext());
                        prefs1.edit().putString(getResources().getString(R.string.quickdicDirectoryKey), t.getText().toString()).commit();
                    });
            b.setNeutralButton(getString(R.string.choose), (dialogInterface, i) -> dirPickerLauncher.launch(null));
            b.create().show();
            return true;
        }
        return super.onPreferenceTreeClick(preference);
    }

    ActivityResultLauncher<Uri> dirPickerLauncher;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String s) {
        final DictionaryApplication application = DictionaryApplication.INSTANCE;
        addPreferencesFromResource(R.xml.preferences);
        ListPreference defaultDic = findPreference(getResources().getString(
                R.string.defaultDicKey));
        List<DictionaryInfo> dicts = application.getDictionariesOnDevice(null);

        final CharSequence[] entries = new CharSequence[dicts.size()];
        final CharSequence[] entryvalues = new CharSequence[dicts.size()];

        for (int i = 0; i < entries.length; ++i) {
            entries[i] = dicts.get(i).dictInfo;
            entryvalues[i] = dicts.get(i).uncompressedFilename;
        }

        defaultDic.setEntries(entries);
        defaultDic.setEntryValues(entryvalues);
        dirPickerLauncher = registerForActivityResult(
                new ActivityResultContracts.OpenDocumentTree() {
                    @Override
                    public Intent createIntent(Context context, Uri input) {
                        Intent intent = super.createIntent(context, input);
                        intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
                        intent.putExtra("android.content.extra.SHOW_ADVANCED", true);
                        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION | Intent.FLAG_GRANT_PREFIX_URI_PERMISSION);
                        return intent;
                    }
                },
                result -> {
                    if (result == null) return;
                    final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
                    getActivity().getContentResolver().takePersistableUriPermission(result, Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
                    prefs.edit().putString(getResources().getString(R.string.quickdicDirectoryKey), result.toString()).commit();
                });
    }
}

File: src/com/hughes/android/dictionary/PreferenceFragment.kt
package com.hughes.android.dictionary

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.InputType
import android.widget.EditText
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts.OpenDocumentTree
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager

class PreferenceFragment : PreferenceFragmentCompat() {
    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        if (preference.key == resources.getString(R.string.quickdicDirectoryKey)) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
            val current: String =
                prefs.getString(resources.getString(R.string.quickdicDirectoryKey), "")!!
            val t = EditText(activity)
            t.setText(current)
            t.inputType = InputType.TYPE_CLASS_TEXT
            val b = AlertDialog.Builder(activity)
                .setTitle(requireActivity().getString(R.string.quickdicDirectoryTitle))
                .setView(t)
                .setNegativeButton(getString(android.R.string.cancel), null)
                .setPositiveButton(
                    getString(android.R.string.ok)
                ) { _, _ ->
                    val prefs1 = PreferenceManager.getDefaultSharedPreferences(requireContext())
                    prefs1.edit().putString(
                        resources.getString(R.string.quickdicDirectoryKey),
                        t.text.toString()
                    ).commit()
                }
            b.setNeutralButton(
                getString(R.string.choose)
            ) { _, _ ->
                dirPickerLauncher!!.launch(null)
            }
            b.create().show()
            return true
        }
        return super.onPreferenceTreeClick(preference)
    }

    var dirPickerLauncher: ActivityResultLauncher<Uri?>? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, s: String?) {
        val application = DictionaryApplication.INSTANCE
        addPreferencesFromResource(R.xml.preferences)
        val defaultDic = findPreference<ListPreference?>(
            resources.getString(
                R.string.defaultDicKey
            )
        )
        val dicts = application.getDictionariesOnDevice(null)

        val entries = arrayOfNulls<CharSequence>(dicts.size + 1)
        val entryvalues = arrayOfNulls<CharSequence>(dicts.size + 1)

        entries[0] = getString(R.string.none)
        entryvalues[0] = null

        for (i in dicts.indices) {
            entries[i + 1] = dicts[i].dictInfo
            entryvalues[i + 1] = dicts[i].uncompressedFilename
        }

        defaultDic!!.entries = entries
        defaultDic.entryValues = entryvalues
        dirPickerLauncher = registerForActivityResult<Uri?, Uri?>(
            object : OpenDocumentTree() {
                override fun createIntent(context: Context, input: Uri?): Intent {
                    val intent = super.createIntent(context, input)
                    intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true)
                    intent.putExtra("android.content.extra.SHOW_ADVANCED", true)
                    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION or Intent.FLAG_GRANT_PREFIX_URI_PERMISSION)
                    return intent
                }
            },
            ActivityResultCallback { result ->
                if (result == null) return@ActivityResultCallback
                val prefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
                requireActivity().contentResolver.takePersistableUriPermission(
                    result,
                    Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                )
                prefs.edit().putString(
                    resources.getString(R.string.quickdicDirectoryKey),
                    result.toString()
                ).commit()
            })
    }
}
